src/Elcodi/Admin/ProductBundle/Form/Type/CategoryType.php

```diff
@@ -82,52 +84,79 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
     {
+        $currentCategoryId = $builder->getData()->getId();
+
         $builder
-            ->add('name', 'text', array(
+            ->add('name', 'text', [
                 'required' => true,
                 'label'    => 'name',
-            ))
-            ->add('slug', 'text', array(
+            ])
+            ->add('slug', 'text', [
                 'required' => false,
                 'label'    => 'slug',
-            ))
-            ->add('root', 'checkbox', array(
+            ])
+            ->add('root', 'checkbox', [
                 'required' => false,
                 'label'    => 'Root Category',
-            ))
-            ->add('enabled', 'checkbox', array(
+            ])
+            ->add('enabled', 'checkbox', [
                 'required' => false,
                 'label'    => 'Visible',
-            ))
-            ->add('metaTitle', 'text', array(
+            ])
+            ->add('metaTitle', 'text', [
                 'required' => false,
                 'label'    => 'metaTitle',
-            ))
-            ->add('metaDescription', 'text', array(
+            ])
+            ->add('metaDescription', 'text', [
                 'required' => false,
                 'label'    => 'metaDescription',
-            ))
-            ->add('metaKeywords', 'text', array(
+            ])
+            ->add('metaKeywords', 'text', [
                 'required' => false,
                 'label'    => 'metaKeywords',
-            ))
-            ->add('parent', 'entity', array(
-                'class'    => $this->categoryFactory->getEntityNamespace(),
-                'required' => false,
-                'label'    => 'parent',
-                'multiple' => false,
-            ))
-            ->add('products', 'entity', array(
+            ])
+            ->add('parent', 'entity', [
+                'class'         => $this->categoryFactory->getEntityNamespace(),
+                'query_builder' => $this->getAvailableCategories($currentCategoryId),
+                'required'      => true,
+                'label'         => 'parent',
+                'multiple'      => false,
+            ])
+            ->add('products', 'entity', [
                 'class'    => $this->productFactory->getEntityNamespace(),
                 'required' => false,
                 'label'    => false,
                 'multiple' => true,
-            ));
+            ]);
 
         $builder->addEventSubscriber($this->getEntityTranslatorFormEventListener());
     }
 
     /**
+     * This method returns a closure used to show only the valid categories to be selected as parent.
+     *
+     * @param integer|null $currentCategoryId The current category id
+     *
+     * @return callable
+     */
+    protected function getAvailableCategories($currentCategoryId)
+    {
+        return function (CategoryRepository $categoryRepository) use ($currentCategoryId) {
+            $queryBuilder = $categoryRepository
```

Comment:
Well the thing is that we have to return a closure where a repository is injected, I think that's something really coupled with the form.